File: src/csharp/examples/random_forest/DataModel.cs

using System;
using System.Collections.Generic;
using System.Configuration.Internal;
using System.IO;
using System.Runtime.InteropServices;
using Alea.IL.CompilerServices;

namespace Tutorial.Cs.examples.RandomForest.DataModel
{
    //using Label = Int32;
    //using FeatureValue = Double;
    using Sample = List<double>;
    using Domain = List<double>;
    using Domains = List<List<double>>;
    using Labels = List<int>;
    using LabeledSample = Tuple<List<double>, int>;
    using LabelHistogram = Tuple<List<int>, int>;
    using Weights = List<int>;
    using Indices = List<int>;

    public class Split
    {
        public int Feature;
        public double Threshold;

        public Split(int feature, double threshold)
        {
            Feature = feature;
            Threshold = threshold;
        }

        public override string ToString()
        {
            return String.Concat("( F: ", this.Feature.ToString(), ", T: ", this.Threshold.ToString(), ")");
        }
    }

    //http://stackoverflow.com/questions/66893/tree-data-structure-in-c-sharp
    public class Tree
    {
        private int label;
        private Split split;
        private Tree low;
        private Tree high;

        //private LinkedList<NTree<T>> children;

        //public NTree(T data)
        //{
        //    data = data;
        //    children = new LinkedList<NTree<T>>();
        //}

        //public void AddChild(T data)
        //{
        //    children.AddFirst(new NTree<T>(data));
        //}

        //public NTree<T> GetChild(int i)
        //{
        //    foreach (NTree<T> n in children)
        //        if (--i == 0)
        //            return n;
        //    return null;
        //}

        //public void Traverse(NTree<T> node, TreeVisitor<T> visitor)
        //{
        //    visitor(node.data);
        //    foreach (NTree<T> kid in node.children)
        //        Traverse(kid, visitor);
        //}
        public override string ToString()
        {
            return String.Concat("( ", low.ToString(), ", ", split.ToString(), ", ", high.ToString(), ")");
        }
    }

    public class Model
    {
        private List<Tree> trees;
        private int numClasses;
    }

}

File: src/fsharp/examples/random_forest/DataModel.fs
(*** hide ***)
module Tutorial.Fs.examples.RandomForest.DataModel

type Label = int

type FeatureValue = float

(**
Array of `FeatureValue`s for a given sample,
e.g.:

    [| sepalLength; sepalWidth; petalLength; petalWidth |]

*)
type Sample = FeatureValue[]

(**
Array of `FeatureValue`s for a given feature but many samples,
e.g:

    [| sepalLength_1; sepalLength_2; ...; sepalLength_n |]

for n samples.
*)
type FeatureArray = FeatureValue[]

(**
Array of `FeatureArray`s,
e.g:

    |                                                                  |
    | |             |  |            |  |             |  |            | |
    | |sepalLength_1|  |sepalWidth_1|  |petalLength_1|  |sepalWidth_1| |
    | |sepalLength_2|  |sepalWidth_2|  |petalLength_2|  |sepalWidth_2| |
    | | ...         |  | ...        |  | ...         |  | ...        | |
    | |sepalLength_m|  |sepalWidth_m|  |petalLength_m|  |sepalWidth_m| |
    | |             |; |            |; |             |; |            | |
    |                                                                  |

m samples with 4 features each.
*)
type FeatureArrays = FeatureArray[]

type Labels = Label[]

(**
Sample with attached Label,
e.g:

    ([| sepalLength; sepalWidth; petalLength; petalWidth |], 1)

a sample with label 1.
*)
type LabeledSample = Sample * Label

type LabelHistogram = int[] * int // counts per class and sum

type Weights = int[]

type Indices = int[]

type Split =
    { Feature : int
      Threshold : FeatureValue }

    static member Create feature threshold =
        { Feature = feature
          Threshold = threshold }

    override this.ToString() = sprintf "{F:%d, T:%.3f}" this.Feature this.Threshold

type Tree =
    | Leaf of Label
    | Node of low : Tree * split : Split * high : Tree
    override this.ToString() =
        match this with
        | Leaf label -> sprintf "%s" (label.ToString())
        | Node(low, split, high) -> sprintf "(%s %s %s)" (low.ToString()) (split.ToString()) (high.ToString())

type Model =
    | RandomForest of trees : Tree[] * numClasses : int

(** 
We abstract from the random number generator, demanding only a function taking an integer `l` and returning a 
random integer between 0 and `l`. The method `getRngFunction` is a factory providing such a function using `System.Random`.
*)
let getRngFunction seed = 
    let rng = System.Random(seed)
    (fun l -> rng.Next(l))